### PR TITLE
Autocomplete: add position property

### PR DESCRIPTION
**Pull request by retrofox**
## Description
The changes suggested in this PR allow setting the position to the autocomplete Popover. 

## How has this been tested?

* It doesn't introduce any change by default.
* Create an Autocomplete and defines the position:

```es6
{
	// ...
	position: 'bottom',
	// ...
},
```

```es6
const autocompleters = [
	{
		name: 'fruit',
		// The prefix that triggers this completer
		triggerPrefix: '~',
		// The option data
		options: [
			{ visual: '🍎', name: 'Apple', id: 1 },
			{ visual: '🍊', name: 'Orange', id: 2 },
			{ visual: '🍇', name: 'Grapes', id: 3 },
		],
		// Returns a label for an option like "🍊 Orange"
		getOptionLabel: option => (
			<span>
				<span className="icon" >{ option.visual }</span>{ option.name }
			</span>
		),
		// Declares that options should be matched by their name
		getOptionKeywords: option => [ option.name ],
		// Declares that the Grapes option is disabled
		isOptionDisabled: option => option.name === 'Grapes',
		// Declares completions should be inserted as abbreviations
		getOptionCompletion: option => (
			<abbr title={ option.name }>{ option.visual }</abbr>
		),

		position: 'bottom', // <- position here
	}
];
```

You should see the autocomplete popover below to the text input component.

## Screenshots 

canvas | React devtools
------|--------
image | Screen Shot 2021-04-13 at 10 24 53 AM

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by wangboxian123** [comment]
how to use this？

**Comment by pablinos** [comment]
Should we add a Storybook example for this?

**Comment by retrofox** [comment]
> Should we add a Storybook example for this?

starting to deal whit this here https://github.com/WordPress/gutenberg/pull/28611/commits/b19c2ed0d8bdd19bb4a57cc4cc8e0ab88c46974b

**Comment by gziolo** [comment]
I'm afraid this leaks the internal implementation of the autocomplete making it harder to change how it is structured in the future. Do we really need to be able to pass down all `popoverProps`. What is the actual use case? If the `position` is a concern here, maybe we should cover only that.

**Comment by retrofox** [comment]
Thanks, Greg.

> I'm afraid this leaks the internal implementation of the autocomplete making it harder to change how it is structured in the future.

I tried to follow how the `Dropdown` component deals with the position property. It's doable passing the Popover props using the [popoverProps](https://developer.wordpress.org/block-editor/reference-guides/components/dropdown/#popoverprops).


> Do we really need to be able to pass down all `popoverProps`.

I guess we could pass only a `position` prop instead of the whole props object. Opted by the whole object because that's what the `Dropdown` does. 

> What is the actual use case? If the `position` is a concern here, ...

It was part of a custom implementation, where the designer asked me whether was possible to move the autocomplete dropdown just below the input element. I don't wanna bothering you with the details :-)

> ... maybe we should cover only that.

Yeah, I think that would be great. Simpler, and clearer.

**Comment by gziolo** [comment]
> Thanks, Greg.
> 
> > I'm afraid this leaks the internal implementation of the autocomplete making it harder to change how it is structured in the future.
> 
> I tried to follow how the `Dropdown` component deals with the position property. It's doable passing the Popover props using the [popoverProps](https://developer.wordpress.org/block-editor/reference-guides/components/dropdown/#popoverprops).

Right, we needed that `popoverProps` in the past to unify some of the existing interfaces but in retrospect, it was a really bad idea. The lesson learned is that we should seek ways to use composition more to support more advanced use cases rather than trying to have one gigantic component that covers them all. I'm happy to hear that `position` is enough for now. It also makes sense to offer this type of functionality from the perspective of the `Autocomplete` API 👍🏻

**Comment by retrofox** [comment]
> Right, we needed that `popoverProps` in the past to unify some of the existing interfaces but in retrospect, it was a really bad idea. 

Nice to know.

> The lesson learned is that we should seek ways to use composition more to support more advanced use cases rather than trying to have one gigantic component that covers them all. 

:100:


> I'm happy to hear that `position` is enough for now. It also makes sense to offer this type of functionality from the perspective of the `Autocomplete` API 👍🏻

yeah, agree. Updating...

**Comment by retrofox** [comment]
> I'm happy to hear that `position` is enough for now. It also makes sense to offer this type of functionality from the perspective of the `Autocomplete` API 👍🏻

Rebased, and handled here 43ab167290d9204df337c27e8d9761c4fab15094.

File: packages/components/src/autocomplete/README.md [hunk 1]
```diff
@@ -97,6 +97,14 @@ Whether to apply debouncing for the autocompleter. Set to true to enable debounc
 -   Type: `Boolean`
 -   Required: No
 
+#### position
+
+Use this string to modify the direction in which the popover should open relative to its parent node.
```
Comment (gziolo) [inline packages/components/src/autocomplete/README.md]: What are other options available? It would be great to link to the list that covers them all.

**Review by gziolo — COMMENTED** [review]
That's a good direction. Thank you for addressing my feedback 👍🏻

File: packages/components/src/autocomplete/README.md (hunk 1 above)
Comment (retrofox) [reply packages/components/src/autocomplete/README.md]: We can make the reference to the Popover component.

**Pull request closed without merging**